Configuration cache fingerprint read errors cannot be deserialized by launcher

### Current Behavior

When configuration cache fingerprint read throws `ConfigurationCacheError`, an induced failure happens on the client side:

<details><summary>Client stack</summary>
<pre>
org.gradle.launcher.daemon.client.DaemonConnectionException: Could not receive a message from the daemon.
	at org.gradle.launcher.daemon.client.DaemonClientConnection.receive(DaemonClientConnection.java:88)
	at org.gradle.launcher.daemon.client.DaemonClient.monitorBuild(DaemonClient.java:246)
	at org.gradle.launcher.daemon.client.DaemonClient.executeBuild(DaemonClient.java:209)
	at org.gradle.launcher.daemon.client.DaemonClient.execute(DaemonClient.java:158)
	at org.gradle.launcher.daemon.client.DaemonClient.execute(DaemonClient.java:97)
	at org.gradle.launcher.cli.RunBuildAction.run(RunBuildAction.java:56)
	at org.gradle.internal.Actions$RunnableActionAdapter.execute(Actions.java:167)
	at org.gradle.launcher.cli.DefaultCommandLineActionFactory$ParseAndBuildAction.execute(DefaultCommandLineActionFactory.java:370)
	at org.gradle.launcher.cli.DefaultCommandLineActionFactory$ParseAndBuildAction.execute(DefaultCommandLineActionFactory.java:334)
	at org.gradle.launcher.cli.DebugLoggerWarningAction.execute(DebugLoggerWarningAction.java:74)
	at org.gradle.launcher.cli.DebugLoggerWarningAction.execute(DebugLoggerWarningAction.java:30)
	at org.gradle.launcher.cli.WelcomeMessageAction.execute(WelcomeMessageAction.java:97)
	at org.gradle.launcher.cli.WelcomeMessageAction.execute(WelcomeMessageAction.java:40)
	at org.gradle.launcher.cli.NativeServicesInitializingAction.execute(NativeServicesInitializingAction.java:50)
	at org.gradle.launcher.cli.NativeServicesInitializingAction.execute(NativeServicesInitializingAction.java:27)
	at org.gradle.launcher.cli.ExceptionReportingAction.execute(ExceptionReportingAction.java:41)
	at org.gradle.launcher.cli.ExceptionReportingAction.execute(ExceptionReportingAction.java:26)
	at org.gradle.launcher.cli.DefaultCommandLineActionFactory$WithLogging.execute(DefaultCommandLineActionFactory.java:482)
	at org.gradle.launcher.Main.doAction(Main.java:35)
	at org.gradle.launcher.bootstrap.EntryPoint.run(EntryPoint.java:52)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:569)
	at org.gradle.launcher.bootstrap.ProcessBootstrap.runNoExit(ProcessBootstrap.java:72)
	at org.gradle.launcher.bootstrap.ProcessBootstrap.run(ProcessBootstrap.java:39)
	at org.gradle.launcher.GradleMain.main(GradleMain.java:31)
Caused by: org.gradle.internal.remote.internal.RecoverableMessageIOException: Could not read message from '/127.0.0.1:55064'.
	at org.gradle.internal.remote.internal.inet.SocketConnection.receive(SocketConnection.java:92)
	at org.gradle.launcher.daemon.client.DaemonClientConnection.receive(DaemonClientConnection.java:82)
	... 26 more
Caused by: java.lang.ClassNotFoundException: org.gradle.internal.cc.base.exceptions.ConfigurationCacheError
	at java.base/java.net.URLClassLoader.findClass(URLClassLoader.java:445)
	at java.base/java.lang.ClassLoader.loadClass(ClassLoader.java:592)
	at java.base/java.lang.ClassLoader.loadClass(ClassLoader.java:525)
	at java.base/java.lang.Class.forName0(Native Method)
	at java.base/java.lang.Class.forName(Class.java:467)
	at java.base/java.io.ObjectInputStream.resolveClass(ObjectInputStream.java:773)
	at org.gradle.internal.serialize.ClassLoaderObjectInputStream.resolveClass(ClassLoaderObjectInputStream.java:42)
	at java.base/java.io.ObjectInputStream.readNonProxyDesc(ObjectInputStream.java:2034)
	at java.base/java.io.ObjectInputStream.readClassDesc(ObjectInputStream.java:1898)
	at java.base/java.io.ObjectInputStream.readOrdinaryObject(ObjectInputStream.java:2224)
	at java.base/java.io.ObjectInputStream.readObject0(ObjectInputStream.java:1733)
	at java.base/java.io.ObjectInputStream$FieldValues.<init>(ObjectInputStream.java:2606)
	at java.base/java.io.ObjectInputStream.defaultReadObject(ObjectInputStream.java:647)
	at java.base/java.lang.Throwable.readObject(Throwable.java:912)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:569)
	at java.base/java.io.ObjectStreamClass.invokeReadObject(ObjectStreamClass.java:1100)
	at java.base/java.io.ObjectInputStream.readSerialData(ObjectInputStream.java:2423)
	at java.base/java.io.ObjectInputStream.readOrdinaryObject(ObjectInputStream.java:2257)
	at java.base/java.io.ObjectInputStream.readObject0(ObjectInputStream.java:1733)
	at java.base/java.io.ObjectInputStream.readObject(ObjectInputStream.java:509)
	at java.base/java.io.ObjectInputStream.readObject(ObjectInputStream.java:467)
	at org.gradle.internal.serialize.DefaultSerializer.read(DefaultSerializer.java:47)
	at org.gradle.launcher.daemon.protocol.DaemonMessageSerializer$SuccessSerializer.read(DaemonMessageSerializer.java:180)
	at org.gradle.launcher.daemon.protocol.DaemonMessageSerializer$SuccessSerializer.read(DaemonMessageSerializer.java:125)
	at org.gradle.internal.serialize.DefaultSerializerRegistry$TaggedTypeSerializer.read(DefaultSerializerRegistry.java:164)
	at org.gradle.internal.serialize.Serializers$StatefulSerializerAdapter$1.read(Serializers.java:36)
	at org.gradle.internal.remote.internal.inet.SocketConnection.receive(SocketConnection.java:83)
	... 27 more
</pre>
</details> 

<details><summary>Original exception that failed to traverse</summary>
<pre>
* Exception is:
org.gradle.internal.cc.base.exceptions.ConfigurationCacheError: Configuration cache state could not be cached: field `value` of `org.gradle.internal.Try$Success` bean found in field `value` of `org.gradle.api.internal.provider.DefaultValueSourceProviderFactory$DefaultObtainedValue` bean found in field `obtainedValue` of `org.gradle.internal.cc.impl.fingerprint.ConfigurationCacheFingerprint$ValueSource` bean found in Gradle runtime: The value cannot be decoded properly with 'JavaObjectSerializationCodec'. It may have been written incorrectly or its data is corrupted.
	at org.gradle.internal.cc.impl.problems.AbstractProblemsListener.onError(AbstractProblemsListener.kt:36)
	at org.gradle.internal.serialize.graph.AbstractIsolateContext.onError(Contexts.kt:384)
	at org.gradle.internal.serialize.graph.codecs.BindingsBackedCodec.decode(BindingsBackedCodec.kt:98)
	at org.gradle.internal.serialize.graph.DefaultReadContext.read(Contexts.kt:283)
	at org.gradle.internal.serialize.graph.BeanPropertyExtensionsKt.readPropertyValue(BeanPropertyExtensions.kt:61)
	at org.gradle.internal.serialize.beans.services.BeanPropertyReader.readFieldOf(BeanPropertyReader.kt:93)
	at org.gradle.internal.serialize.beans.services.BeanPropertyReader.readFieldsOf(BeanPropertyReader.kt:73)
	at org.gradle.internal.serialize.beans.services.BeanPropertyReader.readStateOf(BeanPropertyReader.kt:66)
	at org.gradle.internal.serialize.graph.codecs.BeanCodec.readBeanOf(BeanCodec.kt:61)
	at org.gradle.internal.serialize.graph.codecs.BeanCodec.decode(BeanCodec.kt:43)
	at org.gradle.internal.serialize.graph.CombinatorsKt$reentrant$1$decodeLoop$1.invokeSuspend(Combinators.kt:122)
	at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:33)
	at kotlin.coroutines.ContinuationKt.startCoroutine(Continuation.kt:115)
	at org.gradle.internal.serialize.graph.CombinatorsKt$reentrant$1.decodeLoop(Combinators.kt:123)
	at org.gradle.internal.serialize.graph.CombinatorsKt$reentrant$1.decode(Combinators.kt:85)
	at org.gradle.internal.serialize.graph.codecs.BindingsBackedCodec.decode(BindingsBackedCodec.kt:92)
	at org.gradle.internal.serialize.graph.DefaultReadContext.read(Contexts.kt:283)
	at org.gradle.internal.cc.impl.fingerprint.ConfigurationCacheFingerprintChecker.checkBuildScopedFingerprint(ConfigurationCacheFingerprintChecker.kt:76)
	at org.gradle.internal.cc.impl.fingerprint.ConfigurationCacheFingerprintController.checkBuildScopedFingerprint(ConfigurationCacheFingerprintController.kt:355)
	at org.gradle.internal.cc.impl.DefaultConfigurationCache$checkBuildScopedFingerprint$1.invokeSuspend(DefaultConfigurationCache.kt:771)
	at org.gradle.internal.cc.impl.DefaultConfigurationCache$checkBuildScopedFingerprint$1.invoke(DefaultConfigurationCache.kt)
	at org.gradle.internal.cc.impl.DefaultConfigurationCache$checkBuildScopedFingerprint$1.invoke(DefaultConfigurationCache.kt)
	at org.gradle.internal.cc.impl.DefaultConfigurationCache$readFingerprintFile$1.invokeSuspend(DefaultConfigurationCache.kt:790)
	at org.gradle.internal.cc.impl.DefaultConfigurationCache$readFingerprintFile$1.invoke(DefaultConfigurationCache.kt)
	at org.gradle.internal.cc.impl.DefaultConfigurationCache$readFingerprintFile$1.invoke(DefaultConfigurationCache.kt)
	at org.gradle.internal.serialize.graph.CodecKt$readWith$1$1.invokeSuspend(Codec.kt:147)
	at org.gradle.internal.serialize.graph.CodecKt$readWith$1$1.invoke(Codec.kt)
	at org.gradle.internal.serialize.graph.CodecKt$readWith$1$1.invoke(Codec.kt)
	at org.gradle.internal.serialize.graph.RunningKt$runReadOperation$2.invokeSuspend(Running.kt:33)
	at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:33)
	at kotlin.coroutines.ContinuationKt.startCoroutine(Continuation.kt:115)
	at org.gradle.internal.serialize.graph.RunningKt.runToCompletion(Running.kt:54)
	at org.gradle.internal.serialize.graph.RunningKt.runReadOperation(Running.kt:32)
	at org.gradle.internal.serialize.graph.CodecKt.readWith(Codec.kt:146)
	at org.gradle.internal.cc.impl.DefaultConfigurationCacheIO.withReadContextFor(DefaultConfigurationCacheIO.kt:503)
	at org.gradle.internal.cc.impl.DefaultConfigurationCacheIO.withReadContextFor(DefaultConfigurationCacheIO.kt:495)
	at org.gradle.internal.cc.impl.ConfigurationCacheBuildTreeIO$DefaultImpls.withReadContextFor(ConfigurationCacheBuildTreeIO.kt:83)
	at org.gradle.internal.cc.impl.DefaultConfigurationCacheIO.withReadContextFor(DefaultConfigurationCacheIO.kt:94)
	at org.gradle.internal.cc.impl.ConfigurationCacheBuildTreeIO$DefaultImpls.withReadContextFor$default(ConfigurationCacheBuildTreeIO.kt:78)
	at org.gradle.internal.cc.impl.DefaultConfigurationCache.readFingerprintFile(DefaultConfigurationCache.kt:788)
	at org.gradle.internal.cc.impl.DefaultConfigurationCache.checkBuildScopedFingerprint(DefaultConfigurationCache.kt:769)
	at org.gradle.internal.cc.impl.DefaultConfigurationCache.checkFingerprintAgainstLoadedProperties(DefaultConfigurationCache.kt:755)
	at org.gradle.internal.cc.impl.DefaultConfigurationCache.checkFingerprint(DefaultConfigurationCache.kt:741)
	at org.gradle.internal.cc.impl.DefaultConfigurationCache.checkedFingerprint(DefaultConfigurationCache.kt:545)
	at org.gradle.internal.cc.impl.DefaultConfigurationCache.access$checkedFingerprint(DefaultConfigurationCache.kt:78)
	at org.gradle.internal.cc.impl.DefaultConfigurationCache$checkCandidate$1.invoke(DefaultConfigurationCache.kt:536)
	at org.gradle.internal.cc.impl.DefaultConfigurationCache$checkCandidate$1.invoke(DefaultConfigurationCache.kt:535)
	at org.gradle.internal.cc.impl.ConfigurationCacheRepository$StoreImpl$useForStateLoad$2.invoke(ConfigurationCacheRepository.kt:239)
	at org.gradle.internal.cc.impl.ConfigurationCacheRepository$StoreImpl$useForStateLoad$2.invoke(ConfigurationCacheRepository.kt:235)
	at org.gradle.internal.cc.impl.ConfigurationCacheRepository$withExclusiveAccessToCache$1.get(ConfigurationCacheRepository.kt:320)
	at org.gradle.cache.internal.LockOnDemandCrossProcessCacheAccess.withFileLock(LockOnDemandCrossProcessCacheAccess.java:90)
	at org.gradle.cache.internal.DefaultCacheCoordinator.withFileLock(DefaultCacheCoordinator.java:213)
	at org.gradle.cache.internal.DefaultPersistentDirectoryStore.withFileLock(DefaultPersistentDirectoryStore.java:147)
	at org.gradle.cache.internal.DefaultCacheFactory$ReferenceTrackingCache.withFileLock(DefaultCacheFactory.java:203)
	at org.gradle.internal.cc.impl.ConfigurationCacheRepository.withExclusiveAccessToCache(ConfigurationCacheRepository.kt:318)
	at org.gradle.internal.cc.impl.ConfigurationCacheRepository.access$withExclusiveAccessToCache(ConfigurationCacheRepository.kt:54)
	at org.gradle.internal.cc.impl.ConfigurationCacheRepository$StoreImpl.useForStateLoad(ConfigurationCacheRepository.kt:235)
	at org.gradle.internal.cc.impl.DefaultConfigurationCache.checkCandidate(DefaultConfigurationCache.kt:535)
	at org.gradle.internal.cc.impl.DefaultConfigurationCache.searchForValidEntry(DefaultConfigurationCache.kt:467)
	at org.gradle.internal.cc.impl.DefaultConfigurationCache.access$searchForValidEntry(DefaultConfigurationCache.kt:78)
	at org.gradle.internal.cc.impl.DefaultConfigurationCache$checkFingerprint$1.invoke(DefaultConfigurationCache.kt:455)
	at org.gradle.internal.cc.impl.DefaultConfigurationCache$checkFingerprint$1.invoke(DefaultConfigurationCache.kt:453)
	at org.gradle.configurationcache.ConfigurationCacheBuildOperationsKt$withFingerprintCheckOperations$1.call(ConfigurationCacheBuildOperations.kt:136)
	at org.gradle.configurationcache.ConfigurationCacheBuildOperationsKt$withFingerprintCheckOperations$1.call(ConfigurationCacheBuildOperations.kt:130)
	at org.gradle.internal.operations.DefaultBuildOperationRunner$CallableBuildOperationWorker.execute(DefaultBuildOperationRunner.java:210)
	at org.gradle.internal.operations.DefaultBuildOperationRunner$CallableBuildOperationWorker.execute(DefaultBuildOperationRunner.java:205)
	at org.gradle.internal.operations.DefaultBuildOperationRunner$2.execute(DefaultBuildOperationRunner.java:67)
	at org.gradle.internal.operations.DefaultBuildOperationRunner$2.execute(DefaultBuildOperationRunner.java:60)
	at org.gradle.internal.operations.DefaultBuildOperationRunner.execute(DefaultBuildOperationRunner.java:167)
	at org.gradle.internal.operations.DefaultBuildOperationRunner.execute(DefaultBuildOperationRunner.java:60)
	at org.gradle.internal.operations.DefaultBuildOperationRunner.call(DefaultBuildOperationRunner.java:54)
	at org.gradle.configurationcache.ConfigurationCacheBuildOperationsKt.withFingerprintCheckOperations(ConfigurationCacheBuildOperations.kt:130)
	at org.gradle.internal.cc.impl.DefaultConfigurationCache.checkFingerprint(DefaultConfigurationCache.kt:453)
	at org.gradle.internal.cc.impl.DefaultConfigurationCache.determineCacheAction(DefaultConfigurationCache.kt:387)
	at org.gradle.internal.cc.impl.DefaultConfigurationCache.initializeCacheEntry(DefaultConfigurationCache.kt:188)
	at org.gradle.internal.cc.impl.ConfigurationCacheBuildTreeLifecycleControllerFactory.createRootBuildController(ConfigurationCacheBuildTreeLifecycleControllerFactory.kt:67)
	at org.gradle.composite.internal.DefaultRootBuildState.<init>(DefaultRootBuildState.java:77)
	at org.gradle.composite.internal.BuildStateFactory.createRootBuild(BuildStateFactory.java:68)
	at org.gradle.composite.internal.DefaultIncludedBuildRegistry.createRootBuild(DefaultIncludedBuildRegistry.java:88)
	at org.gradle.launcher.exec.RootBuildLifecycleBuildActionExecutor.execute(RootBuildLifecycleBuildActionExecutor.java:53)
	at org.gradle.internal.buildtree.InitDeprecationLoggingActionExecutor.execute(InitDeprecationLoggingActionExecutor.java:62)
	at org.gradle.internal.buildtree.InitProblems.execute(InitProblems.java:36)
	at org.gradle.internal.buildtree.DefaultBuildTreeContext.execute(DefaultBuildTreeContext.java:40)
	at org.gradle.launcher.exec.BuildTreeLifecycleBuildActionExecutor.lambda$execute$0(BuildTreeLifecycleBuildActionExecutor.java:71)
	at org.gradle.internal.buildtree.BuildTreeState.run(BuildTreeState.java:60)
	at org.gradle.launcher.exec.BuildTreeLifecycleBuildActionExecutor.execute(BuildTreeLifecycleBuildActionExecutor.java:71)
	at org.gradle.launcher.exec.RunAsBuildOperationBuildActionExecutor$2.call(RunAsBuildOperationBuildActionExecutor.java:67)
	at org.gradle.launcher.exec.RunAsBuildOperationBuildActionExecutor$2.call(RunAsBuildOperationBuildActionExecutor.java:63)
	at org.gradle.internal.operations.DefaultBuildOperationRunner$CallableBuildOperationWorker.execute(DefaultBuildOperationRunner.java:210)
	at org.gradle.internal.operations.DefaultBuildOperationRunner$CallableBuildOperationWorker.execute(DefaultBuildOperationRunner.java:205)
	at org.gradle.internal.operations.DefaultBuildOperationRunner$2.execute(DefaultBuildOperationRunner.java:67)
	at org.gradle.internal.operations.DefaultBuildOperationRunner$2.execute(DefaultBuildOperationRunner.java:60)
	at org.gradle.internal.operations.DefaultBuildOperationRunner.execute(DefaultBuildOperationRunner.java:167)
	at org.gradle.internal.operations.DefaultBuildOperationRunner.execute(DefaultBuildOperationRunner.java:60)
	at org.gradle.internal.operations.DefaultBuildOperationRunner.call(DefaultBuildOperationRunner.java:54)
	at org.gradle.launcher.exec.RunAsBuildOperationBuildActionExecutor.execute(RunAsBuildOperationBuildActionExecutor.java:63)
	at org.gradle.launcher.exec.RunAsWorkerThreadBuildActionExecutor.lambda$execute$0(RunAsWorkerThreadBuildActionExecutor.java:36)
	at org.gradle.internal.work.DefaultWorkerLeaseService.withLocks(DefaultWorkerLeaseService.java:263)
	at org.gradle.internal.work.DefaultWorkerLeaseService.runAsWorkerThread(DefaultWorkerLeaseService.java:127)
	at org.gradle.launcher.exec.RunAsWorkerThreadBuildActionExecutor.execute(RunAsWorkerThreadBuildActionExecutor.java:36)
	at org.gradle.tooling.internal.provider.continuous.ContinuousBuildActionExecutor.execute(ContinuousBuildActionExecutor.java:110)
	at org.gradle.tooling.internal.provider.SubscribableBuildActionExecutor.execute(SubscribableBuildActionExecutor.java:64)
	at org.gradle.internal.session.DefaultBuildSessionContext.execute(DefaultBuildSessionContext.java:46)
	at org.gradle.internal.buildprocess.execution.BuildSessionLifecycleBuildActionExecutor$ActionImpl.apply(BuildSessionLifecycleBuildActionExecutor.java:92)
	at org.gradle.internal.buildprocess.execution.BuildSessionLifecycleBuildActionExecutor$ActionImpl.apply(BuildSessionLifecycleBuildActionExecutor.java:80)
	at org.gradle.internal.session.BuildSessionState.run(BuildSessionState.java:73)
	at org.gradle.internal.buildprocess.execution.BuildSessionLifecycleBuildActionExecutor.execute(BuildSessionLifecycleBuildActionExecutor.java:62)
	at org.gradle.internal.buildprocess.execution.BuildSessionLifecycleBuildActionExecutor.execute(BuildSessionLifecycleBuildActionExecutor.java:41)
	at org.gradle.internal.buildprocess.execution.StartParamsValidatingActionExecutor.execute(StartParamsValidatingActionExecutor.java:64)
	at org.gradle.internal.buildprocess.execution.StartParamsValidatingActionExecutor.execute(StartParamsValidatingActionExecutor.java:32)
	at org.gradle.internal.buildprocess.execution.SessionFailureReportingActionExecutor.execute(SessionFailureReportingActionExecutor.java:51)
	at org.gradle.internal.buildprocess.execution.SessionFailureReportingActionExecutor.execute(SessionFailureReportingActionExecutor.java:39)
	at org.gradle.internal.buildprocess.execution.SetupLoggingActionExecutor.execute(SetupLoggingActionExecutor.java:47)
	at org.gradle.internal.buildprocess.execution.SetupLoggingActionExecutor.execute(SetupLoggingActionExecutor.java:31)
	at org.gradle.launcher.daemon.server.exec.ExecuteBuild.doBuild(ExecuteBuild.java:70)
	at org.gradle.launcher.daemon.server.exec.BuildCommandOnly.execute(BuildCommandOnly.java:37)
	at org.gradle.launcher.daemon.server.api.DaemonCommandExecution.proceed(DaemonCommandExecution.java:104)
	at org.gradle.launcher.daemon.server.exec.WatchForDisconnection.execute(WatchForDisconnection.java:39)
	at org.gradle.launcher.daemon.server.api.DaemonCommandExecution.proceed(DaemonCommandExecution.java:104)
	at org.gradle.launcher.daemon.server.exec.ResetDeprecationLogger.execute(ResetDeprecationLogger.java:29)
	at org.gradle.launcher.daemon.server.api.DaemonCommandExecution.proceed(DaemonCommandExecution.java:104)
	at org.gradle.launcher.daemon.server.exec.RequestStopIfSingleUsedDaemon.execute(RequestStopIfSingleUsedDaemon.java:35)
	at org.gradle.launcher.daemon.server.api.DaemonCommandExecution.proceed(DaemonCommandExecution.java:104)
	at org.gradle.launcher.daemon.server.exec.ForwardClientInput.lambda$execute$0(ForwardClientInput.java:40)
	at org.gradle.internal.daemon.clientinput.ClientInputForwarder.forwardInput(ClientInputForwarder.java:80)
	at org.gradle.launcher.daemon.server.exec.ForwardClientInput.execute(ForwardClientInput.java:37)
	at org.gradle.launcher.daemon.server.api.DaemonCommandExecution.proceed(DaemonCommandExecution.java:104)
	at org.gradle.launcher.daemon.server.exec.LogAndCheckHealth.execute(LogAndCheckHealth.java:64)
	at org.gradle.launcher.daemon.server.api.DaemonCommandExecution.proceed(DaemonCommandExecution.java:104)
	at org.gradle.launcher.daemon.server.exec.LogToClient.doBuild(LogToClient.java:63)
	at org.gradle.launcher.daemon.server.exec.BuildCommandOnly.execute(BuildCommandOnly.java:37)
	at org.gradle.launcher.daemon.server.api.DaemonCommandExecution.proceed(DaemonCommandExecution.java:104)
	at org.gradle.launcher.daemon.server.exec.EstablishBuildEnvironment.doBuild(EstablishBuildEnvironment.java:84)
	at org.gradle.launcher.daemon.server.exec.BuildCommandOnly.execute(BuildCommandOnly.java:37)
	at org.gradle.launcher.daemon.server.api.DaemonCommandExecution.proceed(DaemonCommandExecution.java:104)
	at org.gradle.launcher.daemon.server.exec.StartBuildOrRespondWithBusy$1.run(StartBuildOrRespondWithBusy.java:52)
	at org.gradle.launcher.daemon.server.DaemonStateCoordinator.lambda$runCommand$0(DaemonStateCoordinator.java:321)
	at org.gradle.internal.concurrent.ExecutorPolicy$CatchAndRecordFailures.onExecute(ExecutorPolicy.java:64)
	at org.gradle.internal.concurrent.AbstractManagedExecutor$1.run(AbstractManagedExecutor.java:48)
Caused by: java.lang.IllegalArgumentException: Tag guard mismatch for JavaObjectSerializationCodec: expected <172><0>, found <0><0>
	... 137 more
</pre>
</details> 

### Expected Behavior

No induced exception happen, only the original is reported.

### Context (optional)

1. Normal build failures are correctly serialized [here](https://github.com/gradle/gradle/blob/b6a2ccc791c4991ec96dd0781a09842d2b12d774/platforms/core-runtime/build-state/src/main/java/org/gradle/internal/buildprocess/execution/BuildSessionLifecycleBuildActionExecutor.java#L93)
2. An exception is converted to `Result` [here](https://github.com/gradle/gradle/blob/3eda2dd72f9e01f26127f80d01047f4161c4b247/platforms/core-runtime/launcher/src/main/java/org/gradle/tooling/internal/provider/ExecuteBuildActionRunner.java#L34)
3. But the conversion code only wraps build execution [here](https://github.com/gradle/gradle/blob/2e1738412275cda3f16d078f0d6cbeb3e1e1fccf/platforms/core-runtime/launcher/src/main/java/org/gradle/launcher/exec/RootBuildLifecycleBuildActionExecutor.java#L54)
4. While fingerprint checker throws from [here](https://github.com/gradle/gradle/blob/2e1738412275cda3f16d078f0d6cbeb3e1e1fccf/platforms/core-runtime/launcher/src/main/java/org/gradle/launcher/exec/RootBuildLifecycleBuildActionExecutor.java#L53)
5. And then the exception is rethrown [here](https://github.com/gradle/gradle/blob/b6a2ccc791c4991ec96dd0781a09842d2b12d774/platforms/core-runtime/build-state/src/main/java/org/gradle/internal/buildprocess/execution/BuildSessionLifecycleBuildActionExecutor.java#L66)
6. The exception is then converted to the failure but without any custom serialization [here](https://github.com/gradle/gradle/blob/5e9f3d5c77ece47c76373524b2c3ca23dcbfe266/platforms/core-runtime/build-state/src/main/java/org/gradle/internal/buildprocess/execution/SessionFailureReportingActionExecutor.java#L65)
7. An we send the exception serialized with just Java Serialization [here](https://github.com/gradle/gradle/blob/106c48d5d0aff40662ce18b2c27339146fe96318/platforms/core-runtime/daemon-protocol/src/main/java/org/gradle/launcher/daemon/protocol/DaemonMessageSerializer.java#L153)

### Self-contained Reproducer Project

https://github.com/gradle/gradle/blob/d960a5bb92a0e7ca033eb3c6a3dea15dbb42b5e1/platforms/core-configuration/configuration-cache/src/integTest/groovy/org/gradle/internal/cc/impl/ConfigurationCacheIntegrityCheckIntegrationTest.groovy#L77

### Gradle version

8.14

### Build scan URL (optional)

https://ge.gradle.org/s/yjzhojexxqnng

### Your Environment (optional)

_No response_

--- a/platforms/core-configuration/configuration-cache/src/integTest/groovy/org/gradle/internal/cc/impl/ConfigurationCacheIntegrityCheckIntegrationTest.groovy
+++ b/platforms/core-configuration/configuration-cache/src/integTest/groovy/org/gradle/internal/cc/impl/ConfigurationCacheIntegrityCheckIntegrationTest.groovy
@@ -78,55 +78,7 @@ class ConfigurationCacheIntegrityCheckIntegrationTest extends AbstractConfigurat
         )
         failureCauseContains("Tag guard mismatch for JavaObjectSerializationCodec:")
 
-        and: "the error raised while storing state is wrapped so the client can deserialize it (contrast with the fingerprint path, issue 32807)"
-        failure.assertThatDescription(not(containsString(DAEMON_COMMUNICATION_FAILURE)))
-    }
-
-    def "integrity checks detect invalid serialization protocol implementation in fingerprint"() {
-        def configurationCache = new ConfigurationCacheFixture(this)
-
-        buildFile """
-            import org.gradle.api.provider.*
-
-            ${brokenSerializable()}
-
-            abstract class MyValueSource implements ValueSource<CustomSerializable, ValueSourceParameters.None> {
-                @Override
-                CustomSerializable obtain() {
-                    return new CustomSerializable("John", 23)
-                }
-            }
-
-            tasks.register("showUser") {
-                def user = providers.of(MyValueSource) {}.get().name
-
-                doLast {
-                    println("Hello, \$user!")
-                }
-            }
-        """
-
-        when:
-        configurationCacheRun("showUser", "-D${INTEGRITY_CHECKS}=true")
-
-        then:
-        // We only read fingerprint when reusing the cache, it isn't part of load-after-store.
-        // Thus the first (store) run succeeds.
-        configurationCache.assertStateStored()
-
-        when:
-        configurationCacheFails("showUser", "-D${INTEGRITY_CHECKS}=true")
-
-        then:
-        failureDescriptionStartsWith('Configuration cache state could not be cached: ' +
-            'field `value` of `org.gradle.internal.Try$Success` bean found in ' +
-            'field `value` of `org.gradle.api.internal.provider.DefaultValueSourceProviderFactory$DefaultObtainedValue` bean found in ' +
-            'class `MyValueSource`' +
-            ": The value cannot be decoded properly with 'JavaObjectSerializationCodec'. " +
-            "It may have been written incorrectly or its data is corrupted.")
-        failureCauseContains("Tag guard mismatch for JavaObjectSerializationCodec:")
-
-        and: "the fingerprint read error is wrapped so the client can deserialize it"
+        and:
         failure.assertThatDescription(not(containsString(DAEMON_COMMUNICATION_FAILURE)))
     }
 
@@ -160,22 +112,30 @@ class ConfigurationCacheIntegrityCheckIntegrationTest extends AbstractConfigurat
         configurationCacheRun("showUser", "-D${INTEGRITY_CHECKS}=true")
 
         then:
+        // We only read fingerprint when reusing the cache, it isn't part of load-after-store.
+        // Thus the first (store) run succeeds.
         configurationCache.assertStateStored()
 
         when:
         configurationCacheFails("showUser", "-D${INTEGRITY_CHECKS}=true")
 
         then:
-        failureDescriptionContains("The value cannot be decoded properly with 'JavaObjectSerializationCodec'.")
+        failureDescriptionContains(
+            'field `value` of `org.gradle.internal.Try$Success` bean found in ' +
+                'field `value` of `org.gradle.api.internal.provider.DefaultValueSourceProviderFactory$DefaultObtainedValue` bean found in ' +
+                'class `MyValueSource`' +
+                ": The value cannot be decoded properly with 'JavaObjectSerializationCodec'. " +
+                "It may have been written incorrectly or its data is corrupted.")
         failureCauseContains("Tag guard mismatch for JavaObjectSerializationCodec:")
 
-        and: "the fingerprint read error is wrapped so the client can deserialize it"
+        and:
         failure.assertThatDescription(not(containsString(DAEMON_COMMUNICATION_FAILURE)))
 
         where:
-        shape                  | valueType | accessor      | extraClasses                                                                                                                                                    | valueExpression
-        "nested in a bean"     | "Wrapper" | ".inner.name" | "class Wrapper implements Serializable { CustomSerializable inner; Wrapper(CustomSerializable inner) { this.inner = inner } }" | "new Wrapper(new CustomSerializable('John', 23))"
-        "an element of a list" | "List"    | "[0].name"    | ""                                                                                                                                                              | "[new CustomSerializable('John', 23)]"
+        shape                  | valueType            | accessor      | extraClasses                                                                                                                  | valueExpression
+        "the value itself"     | "CustomSerializable" | ".name"       | ""                                                                                                                            | "new CustomSerializable('John', 23)"
+        "nested in a bean"     | "Wrapper"            | ".inner.name" | "class Wrapper implements Serializable { CustomSerializable inner; Wrapper(CustomSerializable inner) { this.inner = inner } }" | "new Wrapper(new CustomSerializable('John', 23))"
+        "an element of a list" | "List"               | "[0].name"    | ""                                                                                                                            | "[new CustomSerializable('John', 23)]"
     }
 
     def "custom exception raised while reading the fingerprint is deserializable by the client"() {
@@ -211,18 +171,18 @@ class ConfigurationCacheIntegrityCheckIntegrationTest extends AbstractConfigurat
         """
 
         when:
-        configurationCacheRun("showUser")
+        configurationCacheRun("showUser", "-D${INTEGRITY_CHECKS}=true")
 
         then:
         configurationCache.assertStateStored()
 
         when:
-        configurationCacheFails("showUser")
+        configurationCacheFails("showUser", "-D${INTEGRITY_CHECKS}=true")
 
         then:
         failureCauseContains("failed to read fingerprint value")
 
-        and: "the user-defined exception is wrapped so the client can deserialize it"
+        and:
         failure.assertThatDescription(not(containsString(DAEMON_COMMUNICATION_FAILURE)))
     }
 

--- a/platforms/core-configuration/configuration-cache/src/integTest/groovy/org/gradle/internal/cc/impl/ConfigurationCacheIntegrityCheckIntegrationTest.groovy
+++ b/platforms/core-configuration/configuration-cache/src/integTest/groovy/org/gradle/internal/cc/impl/ConfigurationCacheIntegrityCheckIntegrationTest.groovy
@@ -18,10 +18,13 @@ package org.gradle.internal.cc.impl
 
 import org.gradle.initialization.StartParameterBuildOptions
 import org.gradle.integtests.fixtures.configurationcache.ConfigurationCacheFixture
-import org.gradle.integtests.fixtures.executer.GradleContextualExecuter
-import org.gradle.util.internal.ToBeImplemented
+
+import static org.hamcrest.CoreMatchers.containsString
+import static org.hamcrest.CoreMatchers.not
 
 class ConfigurationCacheIntegrityCheckIntegrationTest extends AbstractConfigurationCacheIntegrationTest {
+
+    static final String DAEMON_COMMUNICATION_FAILURE = "Could not receive a message from the daemon"
     static final String INTEGRITY_CHECKS = StartParameterBuildOptions.ConfigurationCacheIntegrityCheckOption.PROPERTY_NAME
 
     def "enabling integrity check invalidates CC"() {
@@ -74,9 +77,11 @@ class ConfigurationCacheIntegrityCheckIntegrationTest extends AbstractConfigurat
             "The value cannot be decoded properly with 'JavaObjectSerializationCodec'."
         )
         failureCauseContains("Tag guard mismatch for JavaObjectSerializationCodec:")
+
+        and: "the error raised while storing state is wrapped so the client can deserialize it (contrast with the fingerprint path, issue 32807)"
+        failure.assertThatDescription(not(containsString(DAEMON_COMMUNICATION_FAILURE)))
     }
 
-    @ToBeImplemented("https://github.com/gradle/gradle/issues/32807")
     def "integrity checks detect invalid serialization protocol implementation in fingerprint"() {
         def configurationCache = new ConfigurationCacheFixture(this)
 
@@ -121,10 +126,104 @@ class ConfigurationCacheIntegrityCheckIntegrationTest extends AbstractConfigurat
             "It may have been written incorrectly or its data is corrupted.")
         failureCauseContains("Tag guard mismatch for JavaObjectSerializationCodec:")
 
-        if (GradleContextualExecuter.isDaemon()) {
-            // TODO(https://github.com/gradle/gradle/issues/32807): this is an induced failure that shouldn't happen
-            failureDescriptionContains("Could not receive a message from the daemon")
-        }
+        and: "the fingerprint read error is wrapped so the client can deserialize it"
+        failure.assertThatDescription(not(containsString(DAEMON_COMMUNICATION_FAILURE)))
+    }
+
+    def "integrity checks detect invalid serialization protocol implementation in fingerprint when the broken value is #shape"() {
+        def configurationCache = new ConfigurationCacheFixture(this)
+
+        buildFile """
+            import org.gradle.api.provider.*
+
+            ${brokenSerializable()}
+
+            $extraClasses
+
+            abstract class MyValueSource implements ValueSource<$valueType, ValueSourceParameters.None> {
+                @Override
+                $valueType obtain() {
+                    return $valueExpression
+                }
+            }
+
+            tasks.register("showUser") {
+                def user = providers.of(MyValueSource) {}.get()$accessor
+
+                doLast {
+                    println("Hello, \$user!")
+                }
+            }
+        """
+
+        when:
+        configurationCacheRun("showUser", "-D${INTEGRITY_CHECKS}=true")
+
+        then:
+        configurationCache.assertStateStored()
+
+        when:
+        configurationCacheFails("showUser", "-D${INTEGRITY_CHECKS}=true")
+
+        then:
+        failureDescriptionContains("The value cannot be decoded properly with 'JavaObjectSerializationCodec'.")
+        failureCauseContains("Tag guard mismatch for JavaObjectSerializationCodec:")
+
+        and: "the fingerprint read error is wrapped so the client can deserialize it"
+        failure.assertThatDescription(not(containsString(DAEMON_COMMUNICATION_FAILURE)))
+
+        where:
+        shape                  | valueType | accessor      | extraClasses                                                                                                                                                    | valueExpression
+        "nested in a bean"     | "Wrapper" | ".inner.name" | "class Wrapper implements Serializable { CustomSerializable inner; Wrapper(CustomSerializable inner) { this.inner = inner } }" | "new Wrapper(new CustomSerializable('John', 23))"
+        "an element of a list" | "List"    | "[0].name"    | ""                                                                                                                                                              | "[new CustomSerializable('John', 23)]"
+    }
+
+    def "custom exception raised while reading the fingerprint is deserializable by the client"() {
+        def configurationCache = new ConfigurationCacheFixture(this)
+
+        buildFile """
+            import org.gradle.api.provider.*
+
+            class BrokenFingerprintException extends RuntimeException {
+                BrokenFingerprintException(String message) { super(message) }
+            }
+
+            class ThrowsWhenRead implements Serializable {
+                private void readObject(ObjectInputStream input) throws IOException, ClassNotFoundException {
+                    input.defaultReadObject()
+                    throw new BrokenFingerprintException("failed to read fingerprint value")
+                }
+            }
+
+            abstract class MyValueSource implements ValueSource<ThrowsWhenRead, ValueSourceParameters.None> {
+                @Override
+                ThrowsWhenRead obtain() {
+                    return new ThrowsWhenRead()
+                }
+            }
+
+            tasks.register("showUser") {
+                providers.of(MyValueSource) {}.get()
+                doLast {
+                    println("Hello!")
+                }
+            }
+        """
+
+        when:
+        configurationCacheRun("showUser")
+
+        then:
+        configurationCache.assertStateStored()
+
+        when:
+        configurationCacheFails("showUser")
+
+        then:
+        failureCauseContains("failed to read fingerprint value")
+
+        and: "the user-defined exception is wrapped so the client can deserialize it"
+        failure.assertThatDescription(not(containsString(DAEMON_COMMUNICATION_FAILURE)))
     }
 
     def brokenSerializable() {

--- a/platforms/core-runtime/daemon-messaging/src/main/java/org/gradle/launcher/daemon/protocol/DaemonMessageSerializer.java
+++ b/platforms/core-runtime/daemon-messaging/src/main/java/org/gradle/launcher/daemon/protocol/DaemonMessageSerializer.java
@@ -58,7 +58,7 @@ public class DaemonMessageSerializer {
         registry.register(DaemonUnavailable.class, new DaemonUnavailableSerializer());
         registry.register(BuildStarted.class, new BuildStartedSerializer());
         registry.register(Failure.class, new FailureSerializer(throwableSerializer));
-        registry.register(Success.class, new SuccessSerializer());
+        registry.register(Success.class, new SuccessSerializer(throwableSerializer));
         registry.register(Finished.class, new FinishedSerializer());
 
         // Build events
@@ -82,6 +82,11 @@ public class DaemonMessageSerializer {
     private static class SuccessSerializer implements Serializer<Success> {
         private final Serializer<Object> javaSerializer = new DefaultSerializer<>();
         private final Serializer<SerializedPayload> payloadSerializer = new SerializedPayloadSerializer();
+        private final Serializer<Throwable> throwableSerializer;
+
+        SuccessSerializer(Serializer<Throwable> throwableSerializer) {
+            this.throwableSerializer = throwableSerializer;
+        }
 
         @Override
         public void write(Encoder encoder, Success success) throws Exception {
@@ -107,7 +112,7 @@ public class DaemonMessageSerializer {
                 } else {
                     encoder.writeByte((byte) 4);
                     encoder.writeBoolean(result.wasCancelled());
-                    javaSerializer.write(encoder, result.getException());
+                    throwableSerializer.write(encoder, result.getException());
                 }
             } else {
                 // Serialize anything else
@@ -134,7 +139,7 @@ public class DaemonMessageSerializer {
                     return new Success(BuildActionResult.failed(wasCancelled, failure, null));
                 case 4:
                     wasCancelled = decoder.readBoolean();
-                    RuntimeException exception = (RuntimeException) javaSerializer.read(decoder);
+                    RuntimeException exception = (RuntimeException) throwableSerializer.read(decoder);
                     return new Success(BuildActionResult.failed(wasCancelled, null, exception));
                 case 5:
                     return new Success(javaSerializer.read(decoder));

--- a/platforms/core-runtime/daemon-messaging/src/test/groovy/org/gradle/launcher/daemon/protocol/DaemonMessageSerializerTest.groovy
+++ b/platforms/core-runtime/daemon-messaging/src/test/groovy/org/gradle/launcher/daemon/protocol/DaemonMessageSerializerTest.groovy
@@ -128,6 +128,8 @@ class DaemonMessageSerializerTest extends SerializerSpec {
 
     def "serializes a build failure whose exception cannot be reconstructed by the client as a placeholder"() {
         expect:
+        // Exceptions are sent with plain Java serialization, which rejects a non-transient field whose value
+        // doesn't implement Serializable.
         def unserializable = new RuntimeException("broken") {
             def thing = new Object()
         }
@@ -147,6 +149,8 @@ class DaemonMessageSerializerTest extends SerializerSpec {
         result instanceof Failure
         result.value.getClass() == RuntimeException
 
+        // Exceptions are sent with plain Java serialization, which rejects a non-transient field whose value
+        // doesn't implement Serializable.
         def unserializable = new IOException() {
             def thing = new Object()
         }

--- a/platforms/core-runtime/daemon-messaging/src/test/groovy/org/gradle/launcher/daemon/protocol/DaemonMessageSerializerTest.groovy
+++ b/platforms/core-runtime/daemon-messaging/src/test/groovy/org/gradle/launcher/daemon/protocol/DaemonMessageSerializerTest.groovy
@@ -126,6 +126,19 @@ class DaemonMessageSerializerTest extends SerializerSpec {
         result5.value.exception == null
     }
 
+    def "serializes a build failure whose exception cannot be reconstructed by the client as a placeholder"() {
+        expect:
+        def unserializable = new RuntimeException("broken") {
+            def thing = new Object()
+        }
+        def message = new Success(BuildActionResult.failed(unserializable))
+        def result = serialize(message, serializer)
+        result instanceof Success
+        result.value instanceof BuildActionResult
+        result.value.exception instanceof PlaceholderException
+        result.value.exception.message == "broken"
+    }
+
     def "can serialize Failure messages"() {
         expect:
         def failure = new RuntimeException()

